### ISSUE TITLE
Failed to load config from vite.static.config.ts (Bun 1.1.30)

### ISSUE THREAD

**Issue by SallyBlichG** [oven-sh/bun#14411]
### What version of Bun is running?

v1.1.30 (using the bun action for github actions) 

### What platform is your computer?

ubuntu-22.04 - github actions runner

### What steps can reproduce the bug?

I received this error:
```
bun run build:service_name
bun run build:src:service_name -- id=$id && bun run build:copy:service_name -- id=$id
bunx --bun vite build --config vite.static.config.ts -- env=service_name id=1939
failed to load config from /home/runner/work/repo/repo/dir/vite.static.config.ts
error during build:
Error: config must export or return an object.
    at <anonymous> (/home/runner/work/repo/repo/dir/node_modules/vite/dist/node/chunks/dep-52909643.js:66187:[23](https://github.com/org/repo/actions/runs/11235/job/312355#step:5:24))
    at processTicksAndRejections (:7:39)
error: script "build:src:service_name" exited with code 1
error: script "build:service_name" exited with code 1
```

The build is failing currently

### What is the expected behavior?

```
vite v4.5.2 building for production...
transforming...
✓ 1 modules transformed.
vite v4.5.2 building for production...
transforming...
```

### Additional information

It was working in the previous version - v1.1.29

**Comment by Petucci** [comment]
Hi, we just ran into the same issue while running `bun run vite build` (with `v1.1.301). For now we rolled back to `v1.1.29` and it works without issues.

**Comment by chdenat** [comment]
Got the same error on WSL/Ubuntu 24.04.1 LTS using vite with bun run dev

**Comment by Jarred-Sumner** [comment]
We need to add an integration test for `vite build`

**Comment by Jarred-Sumner** [comment]
Not able to reproduce this yet. I tested with hello world Vite + TS + SWC + React, Vite + TS + Vue, and Remix both build and dev.

Can you provide a reproduction?

**Comment by chdenat** [comment]
WHat I can add is :

The error message is :

```
bun run dev
$ bunx --bun vite --mode development
failed to load config from /home/christian/devs/assets/lgs1920/studio/vite.config.js
error when starting dev server:
Error: config must export or return an object.
    at <anonymous> (/home/christian/devs/assets/lgs1920/studio/node_modules/vite/dist/node/chunks/dep-CDnG8rE7.js:66483:17)
    at processTicksAndRejections (:7:39)
error: script "dev" exited with code 1
```

It fails in a async function loadConfigFromFile 

Here's my vite.config.ts which works with 1.1.29.
```
import {defineConfig} from 'vite';
import react from '@vitejs/plugin-react';
import cesium from 'vite-plugin-cesium';

import data from './public/version.json' with { type: "json" };

const version =data.studio

// https://vitejs.dev/config/
export default defineConfig({
    plugins: [react(), cesium(),
        {
            name: 'markdown-loader',
            transform(code, id) {
                if (id.slice(-3) === '.md') {
                    return `export default ${JSON.stringify(code)};`
                }
            },
        },],
    define: {
        global: {},
    },
    build: {
        minify:'terser',
        target: 'esnext',
        chunkSizeWarningLimit:500000,
        outDir:`./dist/${version}`,
        rollupOptions: {
            output: {
                manualChunks(id: string) {
                    if (id.indexOf('node_modules') !== -1) {
                        const basic = id.toString().split('node_modules/')[1];
                        const sub1 = basic.split('/')[0];
                        if (sub1 !== '.pnpm') {
                            return sub1.toString();
                        }
                        const name2 = basic.split('/')[1];
                        return name2.split('@')[name2[0] === '@' ? 1 : 0].toString();
                    }
                }
            }
        }

    },
    resolve: {
        alias: [

            {
                find: '@Utils',
                replacement: Bun.fileURLToPath(new URL('./src/Utils', import.meta.url))
            },
            {
                find: '@Editor',
                replacement: Bun.fileURLToPath(new URL('./src/components/TracksEditor', import.meta.url))
            },
            {
                find: '@Components',
                replacement: Bun.fileURLToPath(new URL('./src/components', import.meta.url))
            },
            {
                find: '@Core',
                replacement: Bun.fileURLToPath(new URL('./src/core', import.meta.url))
            },
        ]
    }
})
````

**Comment by Petucci** [comment]
In case it helps I added also our config/setup which worked with v1.1.29.

 From our side we get the error:
```
 => ERROR [builder 7/7] RUN bun run build --mode=something                                                                                                                                                                                                                                                                            1.1s 
------
 > [builder 7/7] RUN bun run build --mode=something:
0.564 $ vite build --mode=something
1.051 failed to load config from /app/vite.config.js
1.051 error during build:
1.051 Error: config must export or return an object.
1.051     at <anonymous> (/app/node_modules/vite/dist/node/chunks/dep-Cy9twKMn.js:66422:17)
1.051     at processTicksAndRejections (:7:39)
1.060 error: script "build" exited with code 1
```

when running `RUN bun run build --mode=$MODE`
> Note: I triple checked the `$MODE` variable value and it's set correctly

Our vite config:
```
import {defineConfig} from "vite";
import vue from "@vitejs/plugin-vue";
import path from "path";

// https://vitejs.dev/config/
export default defineConfig({
    envDir: "./env",
    plugins: [vue()],
    resolve: {
        alias: {
            "@": path.resolve(__dirname, "./src"),
        },
    },
    server: {
        open: false,
        port: 8800,
    },
});
```

**Comment by Jarred-Sumner** [comment]
@Petucci @chdenat I'm not able to reproduce this with either of those configs. Any chance you could send us a repo to clone? Maybe it only happens with a specific version of Vite or one of the plugins?

**Comment by SallyBlichG** [comment]
This is our `vite.static.config.ts` file:

```
//Node
import path from 'path'
import { nodePolyfills } from 'vite-plugin-node-polyfills'
//Vite + Rollup
import {
  ConfigEnv,
  defineConfig,
  mergeConfig,
  PluginOption,
  UserConfig,
} from 'vite'
import { viteStaticCopy } from 'vite-plugin-static-copy'
import zipPack from 'vite-plugin-zip-pack'
import serviceMultiple from './vite.plugin'
import copyManifest from './vite.manifest.config'

//
import config, {
  getEntries,
  getCopyFiles,
  getFiles,
  PATHS,
  parseArgs,
} from './vite.common'

export const STATIC_PLUGINS = [
  'vite-plugin-static-copy:serve',
  'vite-plugin-static-copy:build',
  'vite-service-plugin-multiple',
  'vite-plugin-zip-pack',
  'vite-plugin-copy-manifest',
]

const getPlugins = (mode: string): PluginOption[] => {
  const entry = getEntries()
  const files = getCopyFiles(entry)
  const staticFiles = files.filter(file => !file.src.includes('manifest.json'))
  const manifestFiles = files.filter(file =>
    file.src.includes('manifest.json')
  )?.[0]
  const entryFiles = getFiles()
  const args = parseArgs(process.argv)

  const plugins = [
    nodePolyfills() as PluginOption,
    viteStaticCopy({
      targets: staticFiles,
    }) as PluginOption,
    serviceMultiple(entry.shortName, entryFiles, !args.copy),
    copyManifest({
      srcDir: manifestFiles.src as string,
      destDir: `${PATHS.dist}/${manifestFiles.dest}/`,
      version: entry.version,
    }),
  ]

  const xPlugins = [
    zipPack({
      inDir: path.join(PATHS.dist, entry.shortName),
      outDir: PATHS.dist,
      outFileName: `${entry.shortName}_${entry.version}.zip`,
    }),
  ]

  return [...plugins, ...envPlugins]
}

export default defineConfig(
  (env: ConfigEnv): UserConfig =>
    mergeConfig(
      config(env),
      defineConfig({
        plugins: getPlugins(env.mode) || [],
        build: {
          emptyOutDir: false,
          reportCompressedSize: false,
          chunkSizeWarningLimit: 2000,
          rollupOptions: {
            output: {
              entryFileNames: chunk => `${chunk.name}.bundle.js`,
              format: 'iife',
            },
          },
        },
      })
    )
)
```

These are the versions of vite and vite plugins that are being used in our repo:
```
vite@4.5.2
vite-plugin-eslint@1.8.1
vite-plugin-node-polyfills@0.21.0
vite-plugin-static-copy@0.17.1
vite-plugin-zip-pack@1.2.2
vite-tsconfig-paths@4.3.1
```

**Comment by chdenat** [comment]
On my side, the vite packages I'm using.
```
vite-plugin-cesium: "^1.2.23",
@vitejs/plugin-react: "^4.3.2",
vite: "^5.4.8",
```

**Comment by LudvigHz** [comment]
Can confirm this.

packages:
```json
{
   "scripts": {
    "dev": "bunx --bun vite",
    "dev:fullscreen": "VITE_DECORATOR_MODE=fullscreen bunx --bun vite",
    "build": "bunx --bun vite build",
    "build:watch": "bunx --bun vite build --watch",
    "build-and-preview": "bunx --bun vite build && bunx --bun vite preview",
    "preview": "bunx --bun vite preview",
    "build:preview": "bunx --bun vite build --config preview.vite.config.ts",
    "test": "vitest"
  },
  "devDependencies": {
    "@types/react": "^18.3.11",
    "@types/react-dom": "^18.3.1",
    "@vitejs/plugin-react": "^4.3.2",
    "autoprefixer": "^10.4.19",
    "concurrently": "^9.0.0",
    "msw": "^2.4.11",
    "postcss": "^8.4.38",
    "postcss-import": "^16.1.0",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "tailwindcss": "^3.4.3",
    "typescript": "^5.6.3",
    "vite": "^5.4.9",
    "vitest": "^2.1.3",
    "vitest-websocket-mock": "^0.4.0"
  },
  "peerDependencies": {
    "react": "^18.2.0",
    "react-dom": "^18.2.0"
  },
  "dependencies": {
    "@navikt/aksel-icons": "^7.1.0",
    "@navikt/ds-css": "^7.1.0",
    "@navikt/ds-react": "^7.1.0",
    "@navikt/ds-tailwind": "^7.1.0",
    "@navikt/ds-tokens": "^7.1.0",
    "@navikt/navspa": "^6.1.3",
    "@tanstack/query-core": "^5.29.0",
    "classnames": "^2.5.1",
    "use-sync-external-store": "^1.2.0",
    "zustand": "^4.5.2"
  }
}
```
In a monorepo where this package.json is inside a sub-package. I specifically tried another repo that doesn't use workspaces and did not have the same issue (With the same version of vite and bun).

This applies to any vite config. E.g. in the project I get this error:

```ts
import { defineConfig } from 'vite';
import react from '@vitejs/plugin-react';

export default defineConfig({
  plugins: [react()],
  build: {
    cssMinify: true,
    minify: true,
    manifest: 'asset-manifest.json',
    rollupOptions: {
      input: 'src/index.ts',
      preserveEntrySignatures: 'exports-only',
      // external: ["react", "react-dom"],
      output: {
        entryFileNames: 'bundle.js',
        // format: "iife",
        inlineDynamicImports: true,
        assetFileNames: '[name].[ext]',
      },
    },
  },
});
```

**Comment by Jarred-Sumner** [comment]
Can you run bun upgrade —canary and see if it still happens?

**Comment by mik3y** [comment]
Issue remains with the current canary build.

I adjusted the Dockerfile in my repro to run `bun upgrade --canary` prior to build. It fails similarly when updating the base docker tag to canary, e.g. `FROM oven/bun:canary`, which currently corresponds to `1.1.31-canary.37+47ff4748b`:

```
 > [8/8] RUN bun --revision && bun vite build:
0.153 1.1.31-canary.37+47ff4748b
0.381 failed to load config from /app/vite.config.ts
0.381 error during build:
0.381 Error: config must export or return an object.
0.381     at <anonymous> (/app/node_modules/vite/dist/node/chunks/dep-Cyk9bIUq.js:66529:17)
0.381     at processTicksAndRejections (:7:39)
0.385 error: "vite" exited with code 1
```

**Comment by mik3y** [comment]
This is a weaker datapoint, but in a different environment (non-Docker, Mac build), I'm not seeing the issue with `1.1.31-canary.14+584a8ceb8`

**Comment by mik3y** [comment]
Upon further examination of the vite error, it is failing because this internal `isObject()` test is failing:

https://github.com/vitejs/vite/blob/0fe95d4a71930cf55acd628efef59e6eae0f77f7/packages/vite/src/node/utils.ts#L377-L379

```ts
export function isObject(value: unknown): value is Record<string, any> {
  return Object.prototype.toString.call(value) === '[object Object]'
}
```

Debugging the code in question, it looks like `vite.config.ts` is being imported/returned as type `[object Module]` in the affected bun versions. (I'm overall ignorant of what vite is doing here / how imports are resolved, etc.)

**Comment by sapphi-red** [comment]
I think this is happening because bun is not compatible enough with Node.

I noticed that this function assigned to `_require.extensions[loaderExt]` is not called when running on bun.
https://github.com/vitejs/vite/blob/5cca4bfd3202c7aea690acf63f60bfe57fa165de/packages/vite/src/node/config.ts#L1718-L1724
That makes the TS file to be transformed by bun instead of the content transformed by esbuild which Vite is expected.
Because [this line](https://github.com/vitejs/vite/blob/5cca4bfd3202c7aea690acf63f60bfe57fa165de/packages/vite/src/node/config.ts#L1729) expects raw to be `module.exports` from a CJS file (I think this is a valid assumption because Vite controls how the code is generated and never expects the code to be generated in a different way) and the bun's new output does not match that, Vite throws the "config must export or return an object" error.

**Comment by dszymon** [comment]
@Jarred-Sumner it works with `"type": "module"` in `package.json`; w/o it the config object is `Module { default: ... }`

**Comment by chdenat** [comment]
> [@Jarred-Sumner](https://github.com/Jarred-Sumner) it works with `"type": "module"` in `package.json`; w/o it the config object is `Module { default: ... }`

Confirmed on my side, type was set to commonjs and setting it to module did the trick! Thanks @sapphi-red!

**Comment by chdenat** [comment]
Oups... It did the trick with vite 5.4.8... It does not work on vite 5.4.9, same error!

**Comment by chdenat** [comment]
Last comment deleted 

What I did after building my app with type = "module", was a `bun update --latest` to force all packages to be on their latest releases. After that my build failed.
I rollbacked vite to 5.4.8 and got some other errors with imports on some other packages I'm using.
Then I rollbacked type from "module" to "commonjs" and it works... :disturbing effect:

**Comment by Jarred-Sumner** [comment]
The fix for this will release in Bun v1.1.32 (which is in a few hours)
